### ISSUE TITLE
Top toolbar RTA (spectrum analyzer)

### ISSUE THREAD

**Issue by unfa** [LMMS/lmms#2214]
The idea is to add a Foobra-like spectrum analyzer to the top panel of LMMS. There's some space there I think it might be useful to quickly check a frequency, a simple 100Hz/1kHz/10kHz grid could also be added to provide precise reference (why not?) so it's less of just a cool looking toy and more of a real tool. Here's a quick mockup:

![top toolbar rta](https://cloud.githubusercontent.com/assets/491704/8895167/6c5b4b16-33d0-11e5-96ad-6b28c4fc46b8.png)

**Comment by midi-pascal** [comment]
@unfa Interesting idea indeed. But... what does this means in term of additional CPU usage?
I like this but I think it should be optional (I work with a Pentium 4 CPU :)

**Comment by unfa** [comment]
Yeah, totally optional.

**Comment by midi-pascal** [comment]
I think something like the master output widget with a "Click to enable" inside would be cool.

**Comment by Wallacoloo** [comment]
I like the idea. Do you think we should keep both the waveform and spectral
displays separately, or should we merge them into one widget where only one
type of output is displayed at a time? Something like:

Click the waveform display when it's off, and it turns on. Click the
waveform display when it's on and it turns into a spectral display. Click
the display a third time and it turns off again. Repeat.

**Comment by unfa** [comment]
I'd like to see both the waveform and the RTA at the same time, so I'd keep the separate. Similarly like with the "oscilloscope" clicking on it could turn it on/off.

**Comment by midi-pascal** [comment]
Since it can be useful to see both at the same time, I would vote for separate displays.

**Comment by Umcaruje** [comment]
I love this idea :+1:

**Comment by budislav** [comment]
This is good idea but I think it is not so good for ui/ux side. Toolbar is not place for this. I think this should be in one window so it can be resizable and with more options. There can be Phase meter and Peak/RSM Level Meter also. This would be very handy when you do mastering. This can be implemented into tab system.
https://github.com/LMMS/lmms/issues/1911

![visualization](https://cloud.githubusercontent.com/assets/3619927/8897307/313adb80-3413-11e5-8225-23f05325d1a3.png)

**Comment by unfa** [comment]
Actually I don't do mastering in LMMS since I have Ardour, but for mixing this still could be useful.
However I didn't want it to take too much space. Maybe it could be a detachable widget that can be docked in the toolbar or maximized on another screen?

**Comment by budislav** [comment]
I don't want to use two DAWs for one thing.  LMMS should be good for mastering also. Still think that toolbar is not place for that and we need something bigger. You can see a data from all channels not only from master. You can dock this tab on any of horizontal sections and even to maximizing on another screen. That is a power of tab system.

![lmms_dualscreens](https://cloud.githubusercontent.com/assets/3619927/9035301/e78a6ed2-39d6-11e5-828f-d67900e49a72.png)

**Comment by michaelgregorius** [comment]
I agree with @budislav. In my opinion the toolbar is not the correct place for such a tool. Making it possible to change the size of the widget and to dock it would give the users much more options and flexibility. Please also keep in mind that high DPI displays will (hopefully) become more and more common in the not so far future and that it's therefore in LMMS' best interest to be able to render as many elements as possible at different sizes. Unfortunately the current state of the code seems to be that it is cluttered with hard coded sizes and layouts. :(

**Comment by budislav** [comment]
@michaelgregorius for that reason everything should be drawn from software. I used only gradients, outlines, shadows, solid colors and transparency in my designs. So developers can easily develop all patterns like knobs, buttons, levels, backgrounds,...

**Comment by tresf** [comment]
Can someone please explain to me what is different between this request and the Spectrum Analyzer we already have?

If there are features missing from ours, shouldn't we just have an enhancement to improve what we already have?

As far as always being on the screen, we don't have a single-window interface yet, but telling it to always be on top works well for now.

![image](https://cloud.githubusercontent.com/assets/6345473/9441903/95d707dc-4a46-11e5-9502-e90eb21e6b14.png)


![image](https://cloud.githubusercontent.com/assets/6345473/9441826/4de36286-4a46-11e5-8d82-f863e7ad1e9c.png)

![image](https://cloud.githubusercontent.com/assets/6345473/9441842/5e878fc2-4a46-11e5-8e32-7bbcb8e8f67e.png)

On a side note, we really have to start applying some practicality and feasibility to these bug reports.  We're flooded every single day with people's `nice-to-have` wish lists, and it clutters our tracker.  We need to focus on attainable goals and improving what we already have.

**Comment by midi-pascal** [comment]
@tresf <blockquote>
<p><quote>We're flooded every single day with people's nice-to-have wish lists, and it clutters our tracker. We need to focus on attainable goals and improving what we already have.
</blockquote>

:+1: :+1: So many bugs left to fix! :smile:

**Comment by michaelgregorius** [comment]
@tresf I agree. While it is nice to have all the user feedback it can sometimes really clutter the tracker. What I currently do is to sort the issues by "Recently updated" but this will also let issues go out of scope quickly.

I think one of the main problems is that GitHub is not really well suited to do some planning and prioritization. There is no backlog that can be used to prioritize new issues and keep track of what's important next. There are milestones of course but they still only allow for very coarse planning: is it planned for the next release or not? The remaining issues more or less remain in a pile of unsorted, non-prioritized issues.

Another "problem" is that it is hard to plan anyway due to LMMS being developed by some loosely coupled developers of which most do this in their free time I assume. On one hand it is a nice thing because as a developer you get to choose what to work on and what interests you. On the other hand this does not necessarily lead to very coherent development. For example as a developer you do not have to work on the issues in the next milestone. This in turn will lead to delayed releases because you might wait until everything in the official milestone is implemented and fixed. But if there are still some open issues left you might never know whether someone is working on them and whether there is still something to be expected to come.

Last but not least, I think that LMMS lacks some sort of leadership. I think that many of you do a great job in taking care of the issues and organizing them but most of the times it seems like a reaction to things. However, if you take a look at some of the successful open source projects you will see that they often have strong, proactive leaders who push their projects into a direction.

Now I end my off topic rant. :smile:

Perhaps it makes sense to have a new issue that's about the organization and the process (there I said the dreaded word :wink:)?

**Comment by tresf** [comment]
> I think that LMMS lacks some sort of leadership. 

Yes, we lack this method of structure.  This is because from a project perspective, we simply can't afford to be choosers in terms of features and fixes at this stage of the game because we have so little help.

For example... Last year, the lead developer made three new instruments rather than spending time improving our existing VST support.  His perspective was noble, as he considered proprietary code to be a 2nd class citizen in the FOSS world and he refused to fix any bug which solely affected a proprietary binary or OS.

The issue with that perspective is that it puts the objectives of an individual before the objectives of a project.  Most leaders wouldn't allow that, but the project simply cannot afford to loose the help.  We don't turn down charitable donations, because without them, we'd have no donations at all.

That said, I think the phased approach some of our newer team is taking is putting us where we need to be and although from a 1,000 feet there appears to be the appearance of no leadership, reading the last 100 commits shows there is a solid direction and great communication channel happening -- perhaps as an undertone -- which is putting us strategically exactly where we need to be to mature as a software and as a project team. :flags:

**Comment by michaelgregorius** [comment]
@tresf Yes, that's what I meant with that everyone can work on whatever he chooses. From a developer's perspective it's very nice but it does not always help the project. For example I think that many free VSTs (Synth1, Charlatan, Tyrell6, etc.) are of better quality than the internal LMMS plugins so it might have helped more to improve support for VSTs instead of adding three more plugins of rather subpar quality (that also add a maintenance burden). If I remember correctly that developer was mostly concerned with the GUI and then turned to DSP development which is not that easy to master, especially with regards to all the problems that might crop up (aliasing, performance, etc.). Unfortunately, most of LMMS internal plugins exhibit heavy aliasing even when being played in rather "normal" musical regions. This also includes the new ones which even advertise their "bandlimited oscillators" which in fact do not seem to be bandlimited (or they are bandlimited to the ultrasonic band :wink:).

However, I don't want to rant too much on this because I have to admit that's also what I do: pick the things that interest me and that I find the time to work on.

Last but not least, only knowing FOSS software and not taking a look at the proprietary "competition" is also a problem in my opinion because this will lead to some things being implemented in some very strange ways while most of the time most other DAWs implement them in a rather common way. In other words: if you don't know some of the industry standards you cannot expect to be on the same level as them.

**Comment by Wallacoloo** [comment]
@tresf @michaelgregorius Holding this discussion here only serves to force readers of this issue to parse through a lot of unrelated conversation. You should move the conversation elsewhere (be it another issue, a mail-list discussion, wiki discussion page?, etc). Be sure to remove the off-topic comments from this page afterward as well.

**Comment by michaelgregorius** [comment]
@Wallacoloo You're right. I have removed my off-topic comments.
